cmd/thanos/compact.go

```diff
@@ -96,134 +99,102 @@ func runCompact(
-						if id != (ulid.ULID{}) {
-							done = false
-						}
-						continue
-					}
+		// After all compactions are done, work down the downsampling backlog.
+		// We run two passes of this to ensure that the 1h downsampling is generated
+		// for 5m downsamplings created in the first run.
+		level.Info(logger).Log("msg", "start first pass of downsampling")
 
-					if compact.IsIssue347Error(err) {
-						err = compact.RepairIssue347(ctx, logger, bkt, err)
-						if err == nil {
-							done = false
-							continue
-						}
-					}
-					return errors.Wrap(err, "compaction")
-				}
-				if done {
-					break
-				}
-			}
-
-			// After all compactions are done, work down the downsampling backlog.
-			// We run two passes of this to ensure that the 1h downsampling is generated
-			// for 5m downsamplings created in the first run.
-			level.Info(logger).Log("msg", "start first pass of downsampling")
+		if err := downsampleBucket(ctx, logger, bkt, downsamplingDir); err != nil {
+			return errors.Wrap(err, "first pass of downsampling failed")
+		}
 
-			if err := downsampleBucket(ctx, logger, bkt, downsamplingDir); err != nil {
-				return errors.Wrap(err, "first pass of downsampling failed")
-			}
+		level.Info(logger).Log("msg", "start second pass of downsampling")
 
-			level.Info(logger).Log("msg", "start second pass of downsampling")
+		if err := downsampleBucket(ctx, logger, bkt, downsamplingDir); err != nil {
+			return errors.Wrap(err, "second pass of downsampling failed")
+		}
 
-			if err := downsampleBucket(ctx, logger, bkt, downsamplingDir); err != nil {
-				return errors.Wrap(err, "second pass of downsampling failed")
+		if retention.Seconds() != 0 {
+			if err := compact.ApplyDefaultRetentionPolicy(ctx, logger, bkt, retention); err != nil {
+				return errors.Wrap(err, "retention failed")
 			}
-
-			level.Info(logger).Log("msg", "compaction iteration done")
-			return nil
 		}
+		level.Info(logger).Log("msg", "compaction, downsampling and optional retention apply iteration done")
+		return nil
+	}
 
-		g.Add(func() error {
-			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+	g.Add(func() error {
+		defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
 
-			if !wait {
-				return f()
-			}
+		if !wait {
+			return f()
+		}
 
-			// --wait=true is specified.
-			return runutil.Repeat(5*time.Minute, ctx.Done(), func() error {
-				err := f()
-				if err != nil {
-					// The HaltError type signals that we hit a critical bug and should block
-					// for investigation.
-					// You should alert on this being halted.
-					if compact.IsHaltError(err) {
-						if haltOnError {
-							level.Error(logger).Log("msg", "critical error detected; halting", "err", err)
-							halted.Set(1)
-							select {}
-						} else {
-							return errors.Wrap(err, "critical error detected")
-						}
+		// --wait=true is specified.
+		return runutil.Repeat(5*time.Minute, ctx.Done(), func() error {
+			if err := f(); err != nil {
```

Comment:
Sorry for changing the mind, but this is clearly bad. Let's revert this and actually filter out `err == nil` first as clearly we do more on error case. ^^

Reply:
@Bplotka I think you were  right there we should indent error flow as per https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow and keep the successful case at minimal level. WDYT?